doc/org-roam.org

```diff
@@ -1380,6 +1201,141 @@ NOTE: Since ripgrep cannot read encrypted files, this function cannot find
+- Instantaneous SQL-like queries on headlines
+  - What are my TODOs, scheduled for X, or due by Y?
+
+All of these functionality is powered by this database abstraction layer. Hence,
+at its core Org-roam's primary goal is to provide a resilient dual
+representation that is cheap to maintain, easy to understand, and is as
+up-to-date as it possibly can. Org-roam also then exposes an API to this
+database abstraction layer for users who would like to perform programmatic
+queries on their Org files.
+
+** Building Extensions and Advanced Customization of Org-roam
+
+Because Org-roam's core functionality is small, it is possible and sometimes
+desirable to build extensions on top of it. These extensions may one or more of
+the following functionalities:
+
+- Access to Org-roam's database
+- Usage/modification of Org-roam's interactive commands
+
+Org-roam provides no guarantees that extensions will continue to function as
+Org-roam evolves, but by following these simple rules, extensions can be made
+robust to local changes in Org-roam.
+
+1. Extensions should not modify the database schema. Any extension that requires
+   the caching of additional data should make a request upstream to Org-roam.
+2. Extensions requiring access to the database should explicitly state support
+   for the database version (~org-roam-db--version~), and only conditionally
+   load when support is available.
+
+*** Accessing the Database
+
+Access to the database is provided singularly by ~org-roam-db-query~, for
+example:
+
+#+begin_src emacs-lisp
+(org-roam-db-query [:select * :from nodes])
+#+end_src
+
+One can refer to the database schema by looking up
+~org-roam-db--table-schemata~. There are multiple helper functions within
+Org-roam that call ~org-roam-db-query~, these are subject to change. To ensure
+that extensions/customizations are robust to change, extensions should only use
+~org-roam-db-query~, and perhaps replicate the SQL query if necessary.
+
+*** Accessing and Modifying Nodes
+
+The node interface is cleanly defined using ~cl-defstruct~. The primary
+method to access nodes is ~org-roam-node-at-point~ and ~org-roam-node-read~:
+
+- Function: org-roam-node-at-point &optional assert
+
+  Return the node at point. If ASSERT, throw an error if there is no node at
+  point.
+
+- Function: org-roam-node-read &optional initial-input filter-fn sort-fn
+  require-match
+
+  Read and return an ~org-roam-node~.
+  INITIAL-INPUT is the initial prompt value.
+  FILTER-FN is a function to filter out nodes.
+  SORT-FN is a function to sort nodes.
+  If REQUIRE-MATCH, require returning a match.
+
+
+Once you obtain the node, you can use the accessors for the node, e.g.
+~org-roam-node-id~ or ~org-roam-node-todo~.
+
+It is possible to define (or override existing) properties on nodes. This is
+simply done using a ~cl-defmethod~ on the ~org-roam-node~ struct:
+
+#+begin_src emacs-lisp
+(cl-defmethod org-roam-namespace ((node org-roam-node))
```

Comment:
I just realized that in the other comment I didn't understand that it's possible to use other prefix for this function 🤦 I think in this example you should not use `org-roam-` prefix, but something else, e.g. `my-org-roam-namespace`. Otherwise people might copy-paste it and risk a name clash. Because with `org-roam` prefix people extending Org-roam *need* to know all reserved/used names.

Reply:
There's no `org-roam-namespace` now, so no worries about that. Also, I meant the `org-roam-node-` prefix, will fix the typo. Because of how it's currently implemented, implementing with the `org-roam-node-` prefix will make it available in the capture templates, which is the most common use-case.

Org-roam doesn't exactly reserve names, think the rule is not to override the existing interactive functions (e.g. `org-roam-node-read`), or accessors. But everything else is okay to override (e.g. `org-roam-node-slug`).

Reply:
> There's no org-roam-namespace now, so no worries about that.

Yes, **now**. And that's the point. You never know what names users are using and you never know what names you will use in the future. Sharing namespace is begging for problems :) 

> Because of how it's currently implemented, implementing with the org-roam-node- prefix will make it available in the capture templates, which is the most common use-case.

Can you use full name of accessor in capture templates? That would solve the issue.

> Org-roam doesn't exactly reserve names

I mean, the way lisp is loaded, users might accidentally override functions defined in `org-roam`. 

> the rule is not to override the existing interactive functions

It is not future proof :)